File: 2018/Thang10-2018/04-10-2018/houses.pas
Uses    Math;

Const   ginp='houses.inp';
        gout='houses.out';
        maxn=1051;
        base=trunc(1e9)+13;
        di:array[1..4] of longint= (-1,1,0,0);
        dj:array[1..4] of longint= (0,0,-1,1);

Var     m,n,sl,res:longint;
        a,id:array[0..maxn,0..maxn] of longint;
        pow:array[0..maxn] of int64;
        hash:array[0..maxn*maxn,1..4] of int64;
        mv:array[0..maxn*maxn,1..4] of longint;

Procedure Dfs(i,j:longint);
var     k:longint;
Begin
        a[i,j]:=0; id[i,j]:=sl;
        mv[sl,1]:=min(mv[sl,1],i); mv[sl,2]:=max(mv[sl,2],i);
        mv[sl,3]:=min(mv[sl,3],j); mv[sl,4]:=max(mv[sl,4],j);
        for k:=1 to 4 do
                if a[i+di[k],j+dj[k]]=1 then dfs(i+di[k],j+dj[k]);
End;

Procedure Cong(var x:int64; y:int64);
Begin
        x:=(x+y) mod base;
End;

Procedure Sort(l,r:longint);
Var     i,j:longint;
        t,tmp:int64;
Begin
        if l>=r then exit;
        i:=l; j:=r;
        t:=hash[random(r-l+1)+l,1];
        repeat
                while hash[i,1]<t do inc(i);
                while hash[j,1]>t do dec(j);
                if i<=j then
                begin
                        tmp:=hash[i,1]; hash[i,1]:=hash[j,1]; hash[j,1]:=tmp;
                        inc(i); dec(j);
                end;
        until i>j;
        sort(l,j); sort(i,r);
End;

Procedure Enter;
Var     i,j:longint;
Begin
        readln(m,n);
        for i:=1 to m do
        begin
                for j:=1 to n do read(a[i,j]);
                readln;
        end;
        sl:=0;
        for i:=1 to m do
                for j:=1 to n do
                        if a[i,j]=1 then
                        begin
                                inc(sl);
                                mv[sl,1]:=m; mv[sl,2]:=1; mv[sl,3]:=n; mv[sl,4]:=1;
                                dfs(i,j);
                        end;
        pow[0]:=1;
        for i:=1 to max(m,n) do
                pow[i]:=(pow[i-1]*maxn) mod base;
End;

Procedure Process;
Var     i,j,v:longint;
Begin
        for i:=1 to m do
        for j:=1 to n do
        begin
                if id[i,j]=0 then continue;
                v:=id[i,j];
                Cong(hash[v,1], pow[i-mv[v,1]]*pow[j-mv[v,3]]);
                Cong(hash[v,2], pow[i-mv[v,1]]*pow[mv[v,4]-j]);
                Cong(hash[v,3], pow[mv[v,2]-i]*pow[mv[v,4]-j]);
                Cong(hash[v,4], pow[mv[v,2]-i]*pow[j-mv[v,3]]);
        end;
        for i:=1 to sl do
                for j:=2 to 4 do
                        hash[i,1]:=min(hash[i,1],hash[i,j]);
        sort(1,sl);
        res:=0; hash[0,1]:=-1;
        for i:=1 to sl do
                if hash[i,1]<>hash[i-1,1] then inc(res);
        write(res);
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Enter;
        Process;
        Close(input); Close(output);
End.

File: 2019/Thang02_2018/17_02_2019/houses.cpp
#include <iostream>
#include <stdio.h>
#include <algorithm>

using namespace std;
///////

const
    int maxn=1e3+10;
const
    long long base=1e9+13;

int m, n, cnt;
int a[maxn][maxn], id[maxn][maxn];
int adr[maxn*maxn][4];
long long pw[maxn];
long long hs[maxn*maxn][4];
long long h[maxn*maxn];
int res;
///////

void enter()
{
    scanf("%d%d", &m, &n);
    for (int i=1; i<=m; i++)
        for (int j=1; j<=n; j++) scanf("%d", &a[i][j]);
    pw[0]=1;
    for (int i=1; i<=max(m,n); i++)
        pw[i]=pw[i-1]*maxn % base;
}
///////

void dfs(int i, int j)
{
    if (a[i][j]==0 || id[i][j]!=0) return;
    id[i][j]=cnt;
    adr[cnt][0]=min(adr[cnt][0], i);
    adr[cnt][1]=max(adr[cnt][1], j);
    adr[cnt][2]=max(adr[cnt][2], i);
    adr[cnt][3]=min(adr[cnt][3], j);
    dfs(i-1, j);
    dfs(i, j-1);
    dfs(i+1, j);
    dfs(i, j+1);
}
///////

void init()
{
    for (int i=1; i<=m; i++)
        for (int j=1; j<=n; j++)
            if (a[i][j]==1 && id[i][j]==0)
            {
                cnt++;
                adr[cnt][0]=i; adr[cnt][1]=j; adr[cnt][2]=i; adr[cnt][3]=j;
                dfs(i, j);
            }
}
///////

void process()
{
    for (int i=1; i<=m; i++)
        for (int j=1; j<=n; j++)
        if (a[i][j]==1)
        {
            int v=id[i][j];
            int i1=adr[v][0], j2=adr[v][1], i2=adr[v][2], j1=adr[v][3];

            hs[v][0]=(hs[v][0]+pw[i-i1]*pw[j-j1]) % base;
            hs[v][1]=(hs[v][1]+pw[i-i1]*pw[j2-j]) % base;
            hs[v][2]=(hs[v][2]+pw[i2-i]*pw[j2-j]) % base;
            hs[v][3]=(hs[v][3]+pw[i2-i]*pw[j-j1]) % base;
        }
    for (int i=1; i<=cnt; i++)
        h[i]=(hs[i][0]*hs[i][1] % base)*(hs[i][2]*hs[i][3] % base) % base;//min(min(hs[i][0], hs[i][1]), min(hs[i][2], hs[i][3]));
    sort(h+1, h+cnt+1);
    h[0]=-1;
    res=0;
    for (int i=1; i<=cnt; i++)
        if (h[i]!=h[i-1]) res++;
    cout << res;
}
///////

int main()
{
    freopen("houses_0.inp", "r", stdin);
    freopen("houses.out", "w", stdout);
    enter();
    init();
    process();
}
